### PR TITLE
Add Python 3.9 testing and support

### PR DESCRIPTION
**Pull request by stefanor**
Python 3.9 dropped gettimeofday configure checks

In [bpo-38068], cPython assumes gettimeofday exists and takes two arguments. That's a reasonable assumption that we can repeat.

See: python/cpython@f1c1903

[bpo-38068]: https://bugs.python.org/issue38068

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -19,10 +19,10 @@ cache:
 
 python:
   - "2.7"
-  - "3.5"
   - "3.6"
   - "3.7"
   - "3.8"
+  - "3.9"
```
Comment (stefanor) [inline .travis.yml:25]: Not available in travis yet, but I expect it will be in a matter of days. So, better this than `3.9-dev`

**Comment by Erotemic** [comment]
LGTM, ping me when 3.9 is on travis so I can verify it builds on CI.

**Comment by stefanor** [comment]
From https://travis-ci.community/t/python-3-9-0-build/10091/13, looks like it should be there now. Closing and re-opening the PR to trigger a travis retry (can't do that through travis UI for a repo I don't own).

**Comment by stefanor** [comment]
And the 3.9 job got stuck in "Waiting to be queued". Let's try that again.

File: line_profiler/timers.c [hunk 2]
```diff
@@ -48,11 +44,7 @@ hpTimer(void)
 {
         struct timeval tv;
         PY_LONG_LONG ret;
-#ifdef GETTIMEOFDAY_NO_TZ
-        gettimeofday(&tv);
```
Comment (Erotemic) [inline line_profiler/timers.c:52]: Should this line still exist for on Python <= 3.8? If not why?

**Comment by Erotemic** [comment]
To fix the CI, please either allow edits from maintainers or apply this patch: 

```diff
diff --git a/setup.py b/setup.py
index e35ce71..0241f19 100755
--- a/setup.py
+++ b/setup.py
@@ -126,6 +126,7 @@ def parse_requirements(fname='requirements.txt', with_version=False):
 def native_mb_python_tag():
     import sys
     import platform
+    from distutils.version import LooseVersion
     major = sys.version_info[0]
     minor = sys.version_info[1]
     ver = '{}{}'.format(major, minor)
@@ -139,8 +140,8 @@ def native_mb_python_tag():
             else:
                 abi = 'm'
         else:
-            if ver == '38':
-                # no abi in 38?
+            if LooseVersion(ver) >= LooseVersion('38'):
+                # no abi in 38+?
                 abi = ''
             else:
                 abi = 'm'
```

File: line_profiler/timers.c (hunk 2 above)
Comment (stefanor) [reply line_profiler/timers.c:52]: I'll quote from the commit message:

> In bpo-38068, cPython assumes gettimeofday exists and takes two arguments. That's a reasonable assumption that we can repeat.

That was a change made in Python 3.9, but the assumption is a general one, not a python-specific one. I think it's quite reasonable to make.

**Comment by stefanor** [comment]
> To fix the CI, please either allow edits from maintainers or apply this patch

Edits were allowed. Not sure why you weren't able to. But, a modified version of that patch applied.

`LooseVersion()` won't have the correct semantics for `310` < `40` (if that were ever to happen), so didn't use that.

**Comment by Erotemic** [comment]
At, I always wondered what that `m` abi flag was. I could never figure out the right google. Thanks for the link to https://github.com/python/cpython/pull/14097 and subsequently https://bugs.python.org/issue36707 

For my own understanding / reference the summary of the change is: Previously, the `m` suffix used to indicate if `PyObject_MALLOC` was defined as `PyObject_Malloc` (i.e. Python was compiled with pymalloc as the memory allocator). When the suffix wasn't there it meant `PyObject_MALLOC` was defined as `PyMem_MALLOC` (i.e. Python was compiled with malloc as the memory allocator). In 3.8+, the choice of pymalloc or malloc is controlled dynamically at runtime by the `PYTHONMALLOC` environ and thus no longer changes the ABI, which means the suffix can be dropped. Good to know. 

Your change looks appropriate.

File: line_profiler/timers.c (hunk 2 above)
Comment (Erotemic) [reply line_profiler/timers.c:52]: So the existence of this was a quirk of necessary over-configuration? We could have always assumed `gettimeofday(*, *)` existed on POSIX and Windows platforms? I suppose its fine if that's the case.

Comment (stefanor) [reply line_profiler/timers.c:52]: Yeah, you can. Python 3.9 does.

Comment (stefanor) [reply line_profiler/timers.c:52]: It's a special case for ancient unixes, I presume.

**Comment by artemisart** [comment]
> @stefanor To fix the CI, please either allow edits from maintainers or apply this patch:
> 
> ```diff
> diff --git a/setup.py b/setup.py
> index e35ce71..0241f19 100755
> --- a/setup.py
> +++ b/setup.py
> @@ -126,6 +126,7 @@ def parse_requirements(fname='requirements.txt', with_version=False):
>  def native_mb_python_tag():
>      import sys
>      import platform
> +    from distutils.version import LooseVersion
>      major = sys.version_info[0]
>      minor = sys.version_info[1]
>      ver = '{}{}'.format(major, minor)
> @@ -139,8 +140,8 @@ def native_mb_python_tag():
>              else:
>                  abi = 'm'
>          else:
> -            if ver == '38':
> -                # no abi in 38?
> +            if LooseVersion(ver) >= LooseVersion('38'):
> +                # no abi in 38+?
>                  abi = ''
>              else:
>                  abi = 'm'
> ```

It would be *much* better to just use `sys.version_info >= (3, 8)`, lexicographic comparison was made for that.

**Comment by artemisart** [comment]
Why drop 3.5 when it's still compatible at no cost ?

**Comment by boundaryfree** [comment]
Why not push it to pypi?

**Comment by Erotemic** [comment]
I will push this to pypi soon, but I do want to address @artemisart's concern. (This is by far the most widely used project I'm maintaining and I want to do my best to be a good custodian).

I pushed up a PR #37 to test to see if 3.5 still works. I don't remember seeing a justification for why 3.5 was dropped, but I'm also not against dropping it as it is now EOL and will eventually increase maintenance costs. If the dashboard comes back and fails for anything that isn't trivially fixable my feeling is simply drop it and push the 3.9 fix to pypi.

**Comment by Erotemic** [comment]
@artemisart It seems like Python 3.5 still works fine with this patch. I'm not sure why @stefanor opted to remove it other than the fact it is EOL. I do want to remove support for it eventually and clean up all the 3.5 specific code so the package becomes simpler, but that won't happen in `3.1.0`, which I will push to pypi shortly.

**Pull request merged**